server/game/core/ability/abilityTargets/TargetResolverBaseClass.ts

```diff
@@ -0,0 +1,55 @@
+import { ITargetResolverBase } from '../../../TargetInterfaces';
+import { AbilityContext } from '../AbilityContext';
+import * as Contract from '../../utils/Contract';
+import { GameSystem } from '../../gameSystem/GameSystem';
+import PlayerOrCardAbility from '../PlayerOrCardAbility';
+import { RelativePlayer } from '../../Constants';
+
+/**
+ * Base class for all target resolvers.
+ */
+export abstract class TargetResolverBaseClass<TProps extends ITargetResolverBase<AbilityContext>> {
+    protected dependentTarget = null;
+    protected dependentCost = null;
+    public constructor(protected name: string, protected properties: TProps, ability: PlayerOrCardAbility) {
+        if (this.properties.dependsOn) {
+            const dependsOnTarget = ability.targetResolvers.find((target) => target.name === this.properties.dependsOn);
+
+            // assert that the target we depend on actually exists
+            Contract.assertNotNullLike(dependsOnTarget);
+
+            dependsOnTarget.dependentTarget = this;
+        }
+    }
+
+    protected canResolve(context) {
+        // if this depends on another target, that will check hasLegalTarget already
+        return !!this.properties.dependsOn || this.hasLegalTarget(context);
+    }
+
+    protected abstract hasLegalTarget(context): boolean;
+
+    protected getGameSystems(context: AbilityContext): GameSystem | GameSystem[] {
+        return this.properties.immediateEffect ? [this.properties.immediateEffect] : [];
+    }
+
+
+    protected abstract getAllLegalTargets(context: AbilityContext): any[];
```

Comment:
This is a personal thing but typically I put the abstract methods at the top right under the constructor so that they're easy to find for anyone implementing a subclass. Won't block on it, just a suggestion.